airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +47,162 @@ public void checkpoint() {
+     * @return the path to the backup directory
+     */
+    private Path performHsqldbDatabaseBackup() throws IOException {
+
+        String timestamp = DateTimeFormatter.ofPattern("yyyyMMddHHmmss").format(Instant.now());
+        Path source = Paths.get(settingsService.getDefaultJDBCPath()).getParent();
+        Path destination = Paths.get(String.format("%s.backup.%s", settingsService.getDefaultJDBCPath(), timestamp));
+
+        LOG.info("Backing up HSQLDB database to {}...", destination.toString());
+        FileUtils.copyDirectory(source.toFile(), destination.toFile());
+        LOG.info("HSQLDB database backup complete.");
+
+        return destination;
+    }
+
+    /**
+     * Perform an in-place database upgrade from HSQLDB 1.x to 2.x.
+     */
+    private void performHsqldbDatabaseUpgrade() throws SQLException {
+
+        LOG.info("Performing HSQLDB database upgrade...");
+
+        // This will upgrade HSQLDB from 1.8 to 2.5 on the first connection.
+        try (Connection conn = DriverManager.getConnection(settingsService.getDefaultJDBCUrl())) {
+            LOG.info("Database connection established. Current version is: {}.", conn.getMetaData().getDatabaseProductVersion());
+            // On upgrade, the official documentation recommends that we
+            // run 'SHUTDOWN SCRIPT' to compact all the database into a
+            // single SQL file.
+            //
+            // In practice, if we don't do that, we did not observe issues
+            // immediately but after the upgrade.
+            LOG.info("Shutting down database...");
+            try (Statement st = conn.createStatement()) {
+                st.execute("SHUTDOWN SCRIPT");
+            }
+        }
+
+        LOG.info("Finished HSQLDB database upgrade. Current version is: {}.", getHsqldbDatabaseVersion());
+    }
+
+    private void upgradeHsqldbDatabaseSafely() {
+        try {
+            performHsqldbDatabaseBackup();
+            performHsqldbDatabaseUpgrade();
+        } catch (Exception e) {
+            LOG.error("Database upgrade failed, please check backups directory. Current version is: {}.", getHsqldbDatabaseVersion(), e);
+        }
+    }
+
+    /**
+     * Shutdown the embedded HSQLDB database. After this has run, the database cannot be accessed again from the same DataSource.
+     */
+    private void shutdownHsqldbDatabase() {
         try {
-            // Properly shutdown the embedded HSQLDB database.
             LOG.debug("Database shutdown in progress...");
             JdbcTemplate jdbcTemplate = getJdbcTemplate();
-            conn = DataSourceUtils.getConnection(jdbcTemplate.getDataSource());
-            conn.setAutoCommit(true);
-            jdbcTemplate.execute("SHUTDOWN");
+            try (Connection conn = DataSourceUtils.getConnection(jdbcTemplate.getDataSource())) {
+                conn.setAutoCommit(true);
+                jdbcTemplate.execute("SHUTDOWN");
+            }
```

Comment:
Not necessary to set autocommit. This isn't in a transaction, and moreover is an atomic command.
```
getJdbcTemplate().execute("SHUTDOWN");
```

Reply:
You're probably right re:autocommit. I'll remove that.

On the other hand, running `SHUTDOWN` with the JDBCTemplate _permanently_ closes the connection, and all connections after that will fail with `SQLException` (I tried...). I don't know how to entirely reload the pool, do you have an idea?

Reply:
Removed.